packages/eslint-plugin/tests/rules/prefer-optional-chain/prefer-optional-chain.test.ts

```diff
@@ -1866,6 +1866,16 @@ describe('hand-crafted cases', () => {
         ],
       },
 
+      // requireNullish
+      {
+        code: `
+          declare const thing1: string | null;
+          thing1 && thing1.toString();
+        `,
+        options: [{ requireNullish: false }],
+        errors: [{ messageId: `preferOptionalChain` }],
+      },
+
```

Comment:
[Testing] I think it would be nice to add both valid and invalid test cases.

It would also be nice to have a richer set of cases for `requireNullish: true` and `requireNullish: false`.

- different operators (`||`, `&&`, etc..)
- count of operands (2, 3, etc...)
- types of operands: nullish/non-nullish, objects with nested properties, maybe even functions...

Reply:
@auvred do you think that in order to provide a thorough test suite, we should run all ~700 existing tests with `requireNullish` set to both `true` and `false`?
If so, do you have a preferred way of duplicating the existing tests? I wasn't sure if it would be a good idea to simply duplicate the contents of the whole test file.

Reply:
> we should run all ~700 existing tests with `requireNullish` set to both `true` and `false`?

Ah, I don't think so :smile:!

The default value for `requireNullish` is `false`, so it turns out that it's pretty well covered, especially in the `packages/eslint-plugin/tests/rules/prefer-optional-chain/base-cases.ts`.

So we can focus on `requireNullish: true` mostly. Here's a list of cases that might be a good starting point I think:

- ```ts
  declare const foo: string | null
  foo && foo.toString()
  ```

- ```ts
  declare const foo: string
  foo && foo.toString()
  ```

- ```ts
  declare const foo: string | null
  foo && foo.toString() && foo.toString()
  ```

- ```ts
  declare const foo: string
  foo && foo.toString() && foo.toString()
  ```

- ```ts
  declare const foo: { bar: string | null | undefined } | null | undefined
  foo && foo.bar && foo.bar.toString()
  ```

- ```ts
  declare const foo: { bar: string }
  foo && foo.bar && foo.bar.toString()
  ```

- ```ts
  declare const foo: { bar: string | null | undefined } | null | undefined
  foo && foo.bar && foo.bar.toString() && foo.bar.toString()
  ```

- ```ts
  declare const foo: { bar: string }
  foo && foo.bar && foo.bar.toString() && foo.bar.toString()
  ```

... and so on! This rule is quite complex in the sense that it is expected to work on any arbitrary chain of logical expressions. So we need to make sure we cover cases where the correct/incorrect part of the logical chain is at the beginning/middle/end of the expression.

But as per https://github.com/typescript-eslint/typescript-eslint/issues/6887, I don't think we can use the automatic test generation approach...

Reply:
Yeah +1, there are >=2 separate things in discussion:

* Handling for different variations of logical expressions & other AST shapes
* Handling for different types

As long as both have sufficient coverage, it's fine that every cross-combination isn't checked. That'd be a lot of redundant checks!